### ISSUE TITLE
[prefer-reduce-type-parameter] fix results in invalid code

### ISSUE THREAD

**Issue by roikoren755** [typescript-eslint/typescript-eslint#3584]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/prefer-reduce-type-parameter": 2
  }
}
```

```TS
declare const arr1: string[];
declare const arr2: string[];

const text = arr1.reduce<string>((acc) => acc, arr1.shift() as string);
```

```JSON
{
  "compilerOptions": {
    "target": "es6",
    "module": "commonjs",
    "allowJs": true,
    "declaration": true,
    "strict": true,
    "noImplicitAny": false,
    "moduleResolution": "node",
    "esModuleInterop": true,
    "preserveSymlinks": true,
    "resolveJsonModule": true
  },
  "exclude": ["**/*.spec.ts", "**/*.spec.tsx", "**/__tests__", "**/dist/**/*"],
  "include": ["src/**/*", "index.ts"]
}
```

**Expected Result**

One of three options.
Either don't report this as an error, since the default value's type is `string | undefined`, and setting the reduce's type parameter as `string` leads to a TS error.

```TS
declare const arr1: string[];
declare const arr2: string[];

const text = arr1.reduce<string>((acc) => acc, arr1.shift() as string);
```

Or remove the type parameter from the reduce, and leave the one at the typecasting.

```TS
declare const arr1: string[];
declare const arr2: string[];

const text = arr1.reduce((acc) => acc, arr1.shift() as string);
```

Or, finally, remove the casting without breaking the code

```TS
declare const arr1: string[];
declare const arr2: string[];

const text = arr1.reduce<string>((acc) => acc, arr1.shift());
```

**Actual Result**

The fix turns the code into the following

```TS
declare const arr1: string[];
declare const arr2: string[];

const text = arr1.reduce<string><string>((acc) => acc, arr1.shift());
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.28.0` |
| `@typescript-eslint/parser`        | `4.28.0` |
| `TypeScript`                       | `4.3.3` |
| `ESLint`                           | `7.29.0` |
| `node`                             | `14.15.5` |
